blocks/filesystem.py

```diff
@@ -290,28 +290,111 @@ def ls(self, path):
                 names = [name for name in names if name.count('/') == path.count('/')]
 
         paths = ['gs://{}/{}'.format(bucket, name) for name in names]
-        return sorted(paths)
+        return sorted(p.rstrip('/') for p in paths)
 
-    def _prefix(self, path):
-        # find the lowest prefix that doesn't include a pattern match
-        splits = path.split('/')
-        accumulate = []
-        while splits:
-            sub = splits.pop(0)
-            if any(x in sub for x in ['*', '?', '[', ']']):
-                break
-            accumulate.append(sub)
-        return '/'.join(accumulate) if accumulate else None
+    def is_dir(self, path):
+        # Check if a path is a directory, locally or on GCS
+        if self.local(path):
+            return os.path.isdir(path)
+        # These two cases are much faster than checking via a call to ls
+        elif path.endswith('/'):
+            return True
+        elif '.' in path:
```

Comment:
in theory can't a directory name include periods?

Reply:
Yes this is true, I added this as a short circuit to significantly speedup some of the checks we do when copying directories with large numbers of files. That speedup is probably not worth the potential bugs, I'll point to the GCSFileSystem implementation for speed.